azafea/event_processors/endless/metrics/v3/model/_base.py

```diff
@@ -0,0 +1,523 @@
+
+        payload = get_variant(payload)
+        payload_type = payload.get_type_string()
+
+        if payload_type != self.__payload_type__:
+            raise WrongPayloadError(f'Metric event {self.__event_uuid__} needs a '
+                                    f'{self.__payload_type__} payload, but got '
+                                    f'{payload} ({payload_type})')
+
+        return self._get_fields_from_payload(payload)
+
+    @staticmethod
+    def _get_fields_from_payload(payload: GLib.Variant) -> Dict[str, Any]:
+        raise NotImplementedError('Implement this method in final event models')  # pragma: no cover
+
+
+class UnknownEvent(MetricEvent):
+    __abstract__ = True
+
+    event_id = Column(postgresql.UUID(as_uuid=True), nullable=False)
+    payload_data = Column(LargeBinary, nullable=False)
+
+    def _parse_payload(self, maybe_payload: GLib.Variant) -> Dict[str, Any]:
+        return {'payload_data': maybe_payload.get_data_as_bytes().get_data()}
+
+
+class InvalidEvent(UnknownEvent):
+    __abstract__ = True
+
+    error = Column(Unicode, nullable=False)
+
+
+class SingularEvent(MetricEvent):
+    __abstract__ = True
+
+    occured_at = Column(DateTime(timezone=True), nullable=False, index=True)
+
+
+class InvalidSingularEvent(SingularEvent, InvalidEvent):
+    __tablename__ = 'invalid_singular_event_v3'
+
+
+class UnknownSingularEvent(SingularEvent, UnknownEvent):
+    __tablename__ = 'unknown_singular_event_v3'
+
+
+class AggregateEvent(MetricEvent):
+    __abstract__ = True
+
+    period = Column(Unicode(length=1), nullable=False)
+    period_start = Column(DateTime(timezone=False), nullable=False, index=True)
+    count = Column(BigInteger, nullable=False)
+
+
+class InvalidAggregateEvent(AggregateEvent, InvalidEvent):
+    __tablename__ = 'invalid_aggregate_event_v3'
+
+
+class UnknownAggregateEvent(AggregateEvent, UnknownEvent):
+    __tablename__ = 'unknown_aggregate_event_v3'
+
+
+@dataclass
+class RequestChannel:
+    image_id: str
+    site: Dict[str, str]
+    dual_boot: bool
+    live_usb: bool
+
+
+@dataclass
+class RequestData:
+    relative_timestamp: int
+    absolute_timestamp: int
+    singulars: Tuple[GLib.Variant, ...]
+    aggregates: Tuple[GLib.Variant, ...]
+    sha512: str
+
+    def asdict(self) -> Dict[str, Any]:
+        return {
+            'relative_timestamp': self.relative_timestamp,
+            'absolute_timestamp': self.absolute_timestamp,
+            'singulars': self.singulars,
+            'aggregates': self.aggregates,
+        }
+
+
+def parse_record(record: bytes) -> Tuple[RequestData, RequestChannel]:
+    # record is an array of bytes, the concatenation of:
+    # - a datetime encoded on 8 bytes (a 64 bits integer timestamp in microseconds)
+    # - the metrics request (a serialized GVariant)
+    # timestamp_bytes, request_bytes = record[:8], record[8:]
+    request_bytes = record[8:]
+
+    # TODO: is this useful?
```

Comment:
I’m not sure about what to do with this timestamp.

Reply:
Isn't it supposed to be combined with the relative + absolute timestamps sent by the client, to calculate how wrong the client's local clock is?

Reply:
In v2, this timestamp was used by Request model, for the field ```received_at``` (who is useless for event creation, it's just an info field about when the request was received) so I think this is useless in v3

Reply:
Except if we want to store when an event is received

Reply:
We’ll include the timestamp algo in a future PR.